### ISSUE TITLE
GraalVM HTTP client problem read-timeout. 

### ISSUE THREAD

**Issue by Casvad** [micronaut-projects/micronaut-core#2400]
Hi im trying to create my first graalvm application with Micronaut but I have some problem with Micronaut HTTP CLIENT.

When I run the application using `java -jar` I got everything okey. but if I try to generate docker image with graalvm I get Read timeout client exception.  

### Steps to Reproduce

1. Download https://github.com/Casvad/graalvm-micronaut
2. Build project `./gradlew assemble` and create docker image  `./docker-build.sh` 
3. Run container from image and consume the next curl which only call some external service and return the response
```curl -X GET \
  http://localhost:8080/conferences/employees-blocking2 \
  -H 'Accept: */*' \
  -H 'Accept-Encoding: gzip, deflate' \
  -H 'Cache-Control: no-cache' \
  -H 'Connection: keep-alive' \
  -H 'Host: localhost:8080' \
  -H 'Postman-Token: 7e6d75b2-d398-4ca9-a26d-505bf73fe1e3,6ae2ee07-6ea8-450d-8274-00c3a4e4fdb0' \
  -H 'User-Agent: PostmanRuntime/7.20.1' \
  -H 'cache-control: no-cache' 
```

### Expected Behaviour

The endpoint returns 200 ok

### Actual Behaviour

The server returns 500 server error and read timeuout from http client 

### Environment Information

Docker 

### Example Application

- https://github.com/Casvad/graalvm-micronaut

**Comment by Casvad** [comment]
maybe this issue is referred to #2398 but im not closing it so you can read this and maybe get something from the example application

**Comment by jameskleeh** [comment]
Duplicate of #2335

**Comment by ilopmar** [comment]
You have a few options to fix it:

#### Keep using GraalVM 19.2.0.1 or 19.2.1

When you start the native-image and send the curl request you will see the following output:

```
$ ./complete 
09:16:28.879 [main] INFO  io.micronaut.runtime.Micronaut - Startup completed in 53ms. Server Running: http://localhost:8080
Hello
WARNING: The sunec native library, required by the SunEC provider, could not be loaded. This library is usually shipped as part of the JDK and can be found under <JAVA_HOME>/jre/lib/<platform>/libsunec.so. It is loaded at run time via System.loadLibrary("sunec"), the first time services from SunEC are accessed. To use this provider's services the java.library.path system property needs to be set accordingly to point to a location that contains libsunec.so. Note that if java.library.path is not set it defaults to the current working directory.
09:16:34.676 [pool-2-thread-2] ERROR i.m.h.s.netty.RoutingInBoundHandler - Unexpected error occurred: Read Timeout
io.micronaut.http.client.exceptions.ReadTimeoutException: Read Timeout
```

The warning message means that `libsunec.so` it's not available so the HTTPS/SSL handshake won't work because of the missing library. More info here: https://github.com/oracle/graal/blob/master/substratevm/JCA-SECURITY-SERVICES.md#native-implementations

To fix it you can do the following:
```
$ echo $JAVA_HOME
/home/ivan/.sdkman/candidates/java/19.2.0.1-grl
$ ./complete-19.2.0.1 -Djava.library.path=$JAVA_HOME/jre/lib/amd64/
```

With that command line the curl request works as expected.

This has been fixed in the profiles but not included in a release yet: https://github.com/micronaut-projects/micronaut-profiles/commit/130cff4122f24215f8f09ed30e1a35bf37869d19


#### Upgrade to GraalVM 19.3.0

You can install it directly using SDKMan: `sdk use java 19.3.0.r8-grl` (or `sdk use java 19.3.0.r11-grl` for Jdk11) and it works as expected (I just tested it). Now GraalVM includes `libsunec.so` so you don't have to do anything special and it will work out of the box. More info https://www.graalvm.org/docs/release-notes/19_3/#native-image

**Comment by Casvad** [comment]
Thanks Ilopmar, It works locally, I have some problems building docker container with the image `frolvlad/alpine-glibc`  

```
FROM gradle:jdk8 as builder
COPY --chown=gradle:gradle . /home/gradle/complete
WORKDIR /home/gradle/complete
RUN ./gradlew clean
RUN ./gradlew assemble -x test
FROM oracle/graalvm-ce:19.3.0 as graalvm
COPY --from=builder /home/gradle/complete/ /home/app/complete
WORKDIR /home/app/complete
RUN gu install native-image
RUN native-image --no-server -cp build/libs/complete-*-all.jar
FROM frolvlad/alpine-glibc
EXPOSE 8080
COPY --from=graalvm /home/app/complete .
ENTRYPOINT ["./complete"]
```

i was getting this error with 19.3.0 `error while loading shared libraries: libstdc++.so.6: cannot open shared object file: No such file or directory` but with version 19.2.1 it compiles fine. 

so I changed `frolvlad/alpine-glibc` to `openjdk:8` and it works perfectly local. 

But now I have another problem. I tried to upload this image in AWS ECS, I generate the image: 

`https://hub.docker.com/repository/docker/casvad/graal-micronaut/tags?page=1` and im getting read timeout in `AWS ECS`

**Comment by Casvad** [comment]
As Jamekleeh said, maybe this is related to #2335 but in this case is not AWS Lambda, instead, AWS ECS

**Comment by ilopmar** [comment]
Regarding the error trying to build the app in Docker with 19.3.0 I've opened this issue: https://github.com/oracle/graal/issues/1891

As you are using GraalVM 19.2.1 to build the app you need to include `libsunec.so`, please see my previous message and try it.

**Comment by ilopmar** [comment]
@Casvad there is a workaround for being able to create the native image with 19.3.0. Please see the Graal issue.

**Comment by Casvad** [comment]
@ilopmar It works if I use 19.3.0 with --static option running native image, but when the application uses http client, the container stops, I don't know exactly why this is happening. 
![image](https://user-images.githubusercontent.com/17624447/69980810-f9000c00-14fe-11ea-9430-1a6a4d9cc847.png)

**Comment by ilopmar** [comment]
@Casvad Yes, I've seen the same behavior. Everything is related to the same issue and at this moment we need to wait until the GraalVM team fixes it.

**Comment by Casvad** [comment]
Thank you so much @ilopmar for your response, I will be waiting for the GraalVM team!. 
Thanks.

**Comment by ilopmar** [comment]
@Casvad I forgot to mention that if you generate the native image **without** `--static` and install `libstdc++` in `frolvlad/alpine-glibc` it works:

```
$ docker run -it --rm -p 8080:8080 -v `pwd`:/app frolvlad/alpine-glibc sh

/ # apk add libstdc++
fetch http://dl-cdn.alpinelinux.org/alpine/v3.10/main/x86_64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.10/community/x86_64/APKINDEX.tar.gz
(1/1) Installing libstdc++ (8.3.0-r0)
Executing glibc-bin-2.30-r0.trigger
/usr/glibc-compat/sbin/ldconfig: /usr/glibc-compat/lib/ld-linux-x86-64.so.2 is not a symbolic link

OK: 18 MiB in 18 packages

/ # ./app/basic-app_19.3.0-java8
./app/basic-app_19.3.0-java8: /usr/lib/libstdc++.so.6: no version information available (required by ./app/basic-app_19.3.0-java8)
./app/basic-app_19.3.0-java8: /usr/lib/libstdc++.so.6: no version information available (required by ./app/basic-app_19.3.0-java8)
16:35:04.785 [main] INFO  io.micronaut.runtime.Micronaut - Startup completed in 66ms. Server Running: http://7c9c3179d872:8080
``` 

And when I send a curl request to a controller that executes an HTTPS call using Micronaut HTTP Client it works.

**Comment by Casvad** [comment]
yes @ilopmar thanks for the tip.
I have a question, this is the same reason why when we build the container on amazon ecs instance it fails with http read-timeout? my project locally works perfectly.

**Comment by willbuck** [comment]
@Casvad I expect the answer there is yes, based on conversations I've had with the AWS folks; similar architecture is used for both Lambda and ECS iirc. 

I was not able to reproduce the failure you experienced (the 500), have you updated the sample app & found a fix since opening this issue? If so, I will close it as fixed by the GraalVM 19.3 integration we've pushed forward.

**Comment by ilopmar** [comment]
@Casvad we have fixed the related issue (https://github.com/micronaut-projects/micronaut-core/issues/2335) and once the build finish (https://github.com/micronaut-projects/micronaut-core/commit/5dfa1c7bc89396065dbf44072b273832c9e286f9/checks?check_suite_id=414123577) you can try with the 1.3.0 snapshot version if you want: https://docs.micronaut.io/latest/guide/index.html#usingsnapshots

Please also make sure you're using GraalVM 19.3.1.
